README.md
# About this module

This module provides a file field which accepts csv file uploads and visualizes their
contents using Recline.js

## INSTALLATION


+ Download the Reline.js library from https://github.com/okfn/recline 
(zip file) and install in 'sites/all/libraries/'.
+ Enable recline module.

## Supported Backends and File Types

This creates grid, graph, and map data previews for CSV and XLS files based off of the following mechanisms.

It first checks to see if the DKAN Datastore module is installed, and if a datastore has been created for the file. If the datastore is available it uses that to visualize the data. This is extremely scalable since it only queries the first 50 rows of the table in the database. It has been tested with files up to 500 GB and a million+ rows.

If the datastore is not available it checks if the file is a CSV. If it is a CSV it tries to load the file into memory. If it takes longer than a second to load the file it instructs the user that the file is too large to preview. This keeps the page from freezing for larger files.

If the file is a XLS it uses the DataProxy services to preview the file since there is currently not a CSV backend for Recline. DataProxy parses the file and returns it as a data object which is previewed.

## Contributing

We are accepting issues in the dkan issue thread only -> https://github.com/nuams/dkan/issues -> Please label your issue as **"component: recline"** after submitting so we can identify problems and feature requests faster.

If you can, please cross reference commits in this repo to the corresponding issue in the dkan issue thread. You can do that easily adding this text:

```
nuams/dkan#issue_id
``` 

to any commit message or comment replacing **issue_id** with the corresponding issue id.


CHANGELOG.txt
7.x-1.12.x
----------
- Make recline previews work on subdir paths.
- Improved check used to determine if bootstrap is available.

7.x-1.11 2016-02-03
------------------
- Remove unneeded fields UI dependency.
- Upgraded included version of Recline.js library.
- Add a check to see if the bootstrap library is already loaded before adding it.

7.x-1.1 2014-06-27
------------------
- Added CHANGELOG.txt
- Change references to nuCivic
- Fix bug with negative lattitude (nuams/dkan#138)


recline.js
/**
 * Drupal-specific JS helper functions and utils. Not to be confused with the
 * Recline library, which should live in your libraries directory.
 */
;(function ($) {

    // Constants.
    var MAX_LABEL_WIDTH = 77;
    var LABEL_MARGIN = 5;

    // Undefined variables.
    var dataset, views, datasetOptions, fileSize, fileType, router;
    var dataExplorerSettings, state, $explorer, dataExplorer, maxSizePreview;
    var datastoreStatus;

    // Create drupal behavior
    Drupal.behaviors.Recline = {
      attach: function (context) {
        $explorer = $('.data-explorer');
        // Local scoped variables.
        Drupal.settings.recline = Drupal.settings.recline || {};
        fileSize = Drupal.settings.recline.fileSize;
        fileType = Drupal.settings.recline.fileType;
        maxSizePreview = Drupal.settings.recline.maxSizePreview;
        datastoreStatus = Drupal.settings.recline.datastoreStatus;

        dataExplorerSettings = {
          grid: Drupal.settings.recline.grid,
          graph: Drupal.settings.recline.graph,
          map: Drupal.settings.recline.map
        };

        // This is the very basic state collection.
        state = recline.View.parseQueryString(decodeURIComponent(window.location.hash));
        if ('#map' in state) {
          state.currentView = 'map';
        } else if ('#graph' in state) {
          state.currentView = 'graph';
        }

        // Init the explorer.
        init();

        // Attach toogle event.
        $('.recline-embed a.embed-link').on('click', function(){
          $(this).parents('.recline-embed').find('.embed-code-wrapper').toggle();
          return false;
        });
      }
    }

    // make Explorer creation / initialization in a function so we can call it
    // again and again
    function createExplorer (dataset, state, settings) {

      // Remove existing data explorer view.
      dataExplorer && dataExplorer.remove();

      var $el = $('<div />');
      $el.appendTo($explorer);

      var views = [];

      if (settings.grid) {
        views.push({
          id: 'grid',
          label: 'Grid',
          view: new recline.View.SlickGrid({
            model: dataset
          })
        });
      }
      if (settings.graph) {
        state.graphOptions = {
          xaxis: {
            tickFormatter: tickFormatter(dataset),
          },
          hooks:{
            processOffset: [processOffset(dataset)],
            bindEvents: [bindEvents],
          }
        };
        views.push({
          id: 'graph',
          label: 'Graph',
          view: new recline.View.Graph({
            model: dataset,
            state: state
          })
        });
      }
      if (settings.map) {
        views.push({
          id: 'map',
          label: 'Map',
          view: new recline.View.Map({
            model: dataset
          })
        });
      }

      // Multiview settings
      var multiviewOptions = {
        model: dataset,
        el: $el,
        state: state,
        views: views
      };

      // Getting base embed url.
      var urlBaseEmbed = $('.embed-code').text();
      var iframeOptions = {src: urlBaseEmbed, width:850, height:400};

      // Attaching router to dataexplorer state.
      dataExplorer = new recline.View.MultiView(multiviewOptions);
      router = new recline.DeepLink.Router(dataExplorer);

      // Adding router listeners.
      var changeEmbedCode = getEmbedCode(iframeOptions);
      router.on('init', changeEmbedCode);
      router.on('stateChange', changeEmbedCode);

      // Add map dependency just for map views.
      _.each(dataExplorer.pageViews, function(item, index){
        if(item.id && item.id === 'map'){
          var map = dataExplorer.pageViews[index].view.map;
          router.addDependency(new recline.DeepLink.Deps.Map(map, router));
        }
      });

      // Start to track state chages.
      router.start();

      $.event.trigger('createDataExplorer');
      return views;
    }

    // Returns the dataset configuration.
    function getDatasetOptions () {
      var datasetOptions = {};
      var delimiter = Drupal.settings.recline.delimiter;
      var file = Drupal.settings.recline.file;
      var uuid = Drupal.settings.recline.uuid;

     // Get correct file location, make sure not local
     file = (getOrigin(window.location) !== getOrigin(file)) ? '/node/' + Drupal.settings.recline.uuid + '/data' : file;

      // Select the backend to use
      switch(getBackend(datastoreStatus, fileType)) {
        case 'csv':
          datasetOptions = {
            backend: 'csv',
            url: file,
            delimiter: delimiter
          };
          break;
        case 'ckan':
          datasetOptions = {
            endpoint: 'api',
            id: uuid,
            backend: 'ckan'
          };
          break;
        case 'xls':
          datasetOptions = {
            backend: 'xls',
            url: file
          };
          break;
        case 'dataproxy':
          datasetOptions = {
            url: file,
            backend: 'dataproxy'
          };
          break;
        default:
          showError('File type ' + fileType + ' not supported for preview.');
          break;
      }

      return datasetOptions;
    }

     // Correct for fact that IE does not provide .origin
    function getOrigin(u) {
      var url = parseURL(u);
        return url.protocol + '//' + url.hostname + (url.port ? (':' + url.port) : '');
    }

    // Parse a simple URL string to get its properties
    function parseURL(url) {
      var parser = document.createElement('a');
      parser.href = url;
      return {
        protocol: parser.protocol,
        hostname: parser.hostname,
        port: parser.port,
        pathname: parser.pathname,
        search: parser.search,
        hash: parser.hash,
        host: parser.host
      }
    }

    // Retrieve a backend given a file type and and a datastore status.
    function getBackend (datastoreStatus, fileType) {

      // If it's inside the datastore then we use the dkan API
      if (datastoreStatus) return 'ckan';
      var formats = {
        'csv': ['text/csv', 'csv'],
        'xls': ['application/vnd.ms-excel', 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'],
      };
      var backend = _.findKey(formats, function(format) { return _.include(format, fileType) });

      // If the backend is an xls but the browser version is prior 9 then
      // we need to fallback to dataproxy
      if (backend === 'xls' && document.documentMode < 9) return 'dataproxy';
      return backend;
    }

    // Displays an error retrieved from the response object.
    function showRequestError (response) {

      // Actually dkan doesn't provide standarization over
      // error handling responses. For example: if you request
      // unexistent resources it will retrive an array with a
      // message inside.
      // Recline backends will return an object with an error.
      try {
        var ro = (typeof response === 'string') ? JSON.parse(response) : response;

        if(ro.error)  {
          showError(ro.error.message)
        } else if(ro instanceof Array) {
          showError(ro[0]);
        }
      } catch (error) {
        showError(response);
      }
    }

    // Displays an error.
    function showError (message) {
      $explorer.html('<div class="messages error">' + message + '</div>');
    }

    // Creates the embed code.
    function getEmbedCode (options){
      return function(state){
        var iframeOptions = _.clone(options);
        var iframeTmpl = _.template('<iframe width="<%= width %>" height="<%= height %>" src="<%= src %>" frameborder="0"></iframe>');
        var previewTmpl = _.template('<%= src %>');
        _.extend(iframeOptions, {src: iframeOptions.src + '#' + (state.serializedState || '')});
        var html = iframeTmpl(iframeOptions);
        $('.embed-code').text(html);
        var preview = previewTmpl(iframeOptions);
        $('.preview-code').text(preview);
      };
    }

    // Creates the preview url code.
    function getPreviewCode (options){
      return function(state){
        var previewOptions = _.clone(options);
        var previewTmpl = _.template('<%= src %>');
        _.extend(previewOptions, {src: previewOptions.src + '#' + (state.serializedState || '')});
        var html = previewTmpl(previewOptions);
        $('.preview-url').text(html);
      };
    }

    // Check if a chart has their axis inverted.
    function isInverted (){
      return dataExplorer.pageViews[1].view.state.attributes.graphType === 'bars';
    }

    // Computes the width of a chart.
    function computeWidth (plot, labels) {
      var biggerLabel = '';
      for( var i = 0; i < labels.length; i++){
        if(labels[i].length > biggerLabel.length && !_.isUndefined(labels[i])){
          biggerLabel = labels[i];
        }
      }
      var canvas = plot.getCanvas();
      var ctx = canvas.getContext('2d');
      ctx.font = 'sans-serif smaller';
      return ctx.measureText(biggerLabel).width;
    }

    // Resize a chart.
    function resize (plot) {
      var itemWidth = computeWidth(plot, _.pluck(plot.getXAxes()[0].ticks, 'label'));
      var graph = dataExplorer.pageViews[1];
      if(!isInverted() && $('#prevent-label-overlapping').is(':checked')){
        var canvasWidth = Math.min(itemWidth + LABEL_MARGIN, MAX_LABEL_WIDTH) * plot.getXAxes()[0].ticks.length;
        var canvasContainerWith = $('.panel.graph').parent().width();
        if(canvasWidth < canvasContainerWith){
            canvasWidth = canvasContainerWith;
        }
        $('.panel.graph').width(canvasWidth);
        $('.recline-flot').css({overflow:'auto'});
      }else{
        $('.recline-flot').css({overflow:'hidden'});
        $('.panel.graph').css({width: '100%'});
      }
      plot.resize();
      plot.setupGrid();
      plot.draw();
    }

    // Bind events after chart resizes.
    function bindEvents (plot, eventHolder) {
      var p = plot || dataExplorer.pageViews[1].view.plot;
      resize(p);
      setTimeout(addCheckbox, 0);
    }

    // Compute the chart offset to display ticks properly.
    function processOffset (dataset) {
      return function(plot, offset) {
        if(dataExplorer.pageViews[1].view.xvaluesAreIndex){
          var series = plot.getData();
          for (var i = 0; i < series.length; i++) {
            var numTicks = Math.min(dataset.records.length, 200);
            var ticks = [];
            for (var j = 0; j < dataset.records.length; j++) {
              ticks.push(parseInt(j, 10));
            }
            if(isInverted()){
              series[i].yaxis.options.ticks = ticks;
            }else{
              series[i].xaxis.options.ticks = ticks;
            }
          }
        }
      };
    }

    // Format ticks base on previews computations.
    function tickFormatter (dataset){
      return function (x) {
        x = parseInt(x, 10);
        try {
          if(isInverted()) return x;
          var field = dataExplorer.pageViews[1].view.state.get('group');
          var label = dataset.records.models[x].get(field) || '';
          if(!moment(String(label)).isValid() && !isNaN(parseInt(label, 10))){
            label = parseInt(label, 10) - 1;
          }
          return label;
        } catch(e) {
          return x;
        }
      };
    }

    // Add checkbox to control resize behavior.
    function addCheckbox () {
      $control = $('.form-stacked:visible').find('#prevent-label-overlapping');
      if(!$control.length){
        $form = $('.form-stacked');
        $checkboxDiv = $('<div class="checkbox"></div>').appendTo($form);
        $label = $('<label />', { 'for': 'prevent-label-overlapping', text: 'Resize graph to prevent label overlapping' }).appendTo($checkboxDiv);
        $label.prepend($('<input />', { type: 'checkbox', id: 'prevent-label-overlapping', value: '' }));
        $control = $('#prevent-label-overlapping');
        $control.on('change', function(){
          resize(dataExplorer.pageViews[1].view.plot);
        });
      }
    }

    // Init the multiview.
    function init () {
      if(fileSize < maxSizePreview || datastoreStatus) {
        dataset = new recline.Model.Dataset(getDatasetOptions());
        dataset.fetch().fail(showRequestError);
        views = createExplorer(dataset, state, dataExplorerSettings);
        views.forEach(function(view) { view.id === 'map' && view.view.redraw('refresh') });
      } else {
        showError('File was too large or unavailable for preview.');
      }
    }

})(jQuery);


recline.api.php
<?php

/**
 * Adds column to recline field for storing non-view data.
 *
 *  @param array columns
 *  Array of existing schema definitions.
 *
 */
function hook_recline_field_columns(&$columns) {
  $columns['my_new_view_additional_data'] = array(
    'type' => 'varchar',
    'length' => '32',
    'description' => 'Column to determine the choropleth map data.',
  );
}


/**
 * Adds an additional view to Recline view options.
 *
 * @return
 * A key/value of the new view.
 */
function hook_recline_view_options() {
  // Adds "my_new_view" column to database and "My New View" to list of Recline
  // views.
  return array('my_new_view' => t('My New View'));
}

/**
 * Additional columns and views need to be added to existing fields.
 * Below is an example using the addtional column and view defined above.
 * my_new_view_additional_data() is schema defined in
 * hook_recline_field_columns.
 */
function hook_update_N(&$sandbox) {
  $ret = array();
  $fields = field_info_fields();
  foreach ($fields as $field_name => $field) {
    if ($field['type'] == 'recline_field' && $field['storage']['type'] == 'field_sql_storage') {
      foreach ($field['storage']['details']['sql'] as $type => $table_info) {
        foreach ($table_info as $table_name => $columns) {
          $column_name = _field_sql_storage_columnname($field_name, 'my_new_view_additional_data');
          // Adding my_new_view_additional_data.
          if (!(db_field_exists($table_name, $column_name))) {
            // Calling schema defined in hook_recline_field_column().
            $schema = my_new_view_additional_data();
            db_add_field($table_name, $column_name, $schema);
          }
          // Adding my_new_view.
          $column_name = _field_sql_storage_columnname($field_name, 'my_new_view');
          $schema = recline_field_schema();
          if (!(db_field_exists($table_name, $column_name))) {
            db_add_field($table_name, $column_name, $schema['columns']['my_new_view']);
          }
          field_cache_clear();
        }
      }
    }
  }
  return $ret;
}

/**
 * The following are examples of adding the additional column data to the
 * recline widget.
 */

/**
 * Implements hook_field_widget_form_alter().
 */
function hook_field_widget_form_alter(&$element, &$form_state, $context) {
  if ($context['field']['type'] == 'recline_field') {
    foreach ($element as $delta => $instance) {
      $my_new_view_additional_data = isset($element[$delta]['#default_value']['my_new_view_additional_data']) ? $element[$delta]['#default_value']['my_new_view_additional_data'] : FALSE;
      $element[$delta]['my_new_view_additional_data'] = array(
        '#title' => 'My Additional Data',
        '#description' => t('Enter additional data.'),
        '#type' => 'textfield',
        '#weight' => 1,
        '#default_value' => $my_new_view_additional_data,
      );
    }
  }
}

/**
 * The following is an example of responding to a new view or additional column.
 * The implementing module is responsible for adding additional view to page.
 */

/**
 * Implements hook_theme_registry_alter().
 */
function hook_theme_registry_alter(&$theme_registry) {
  $theme_registry['recline_default_formatter']['function'] = 'MY_MODULE_recline_default_formatter';
}

/**
 * Adds js settings from recline field.
 */
function MY_MODULE_recline_default_formatter($variables) {
  $output = recline_preview_csv($variables);
  $settings = array();
  if (isset($variables['item']['my_new_view_additional_data']) || isset($variables['item']['my_new_view_additional_data'])) {
    $settings['choropleth'] = array(
      'my_new_view_additional_data' => $variables['item']['my_new_view_additional_data'],
      'my_new_view' => $variables['item']['my_new_view'],
    );
    drupal_add_js($settings, 'setting');
  }
  drupal_add_js('path/to/new_recline_view.js');
  return drupal_render($output);
}


dkan-module-init.sh

# Name of the current module.
DKAN_MODULE="recline"

# DKAN branch to use
DKAN_BRANCH="release-1-12"

COMPOSER_PATH=".composer/vendor/bin"

#Fix for probo not setting the composer path.
if [[ "$PATH" != *"$COMPOSER_PATH"* ]]; then
  echo "> Composer PATH is not set. Adding temporarily.. (you should add to your .bashrc)"
  echo "PATH (prior) = $PATH"
  export PATH="$PATH:~/$COMPOSER_PATH"
fi

wget -O /tmp/dkan-init.sh https://raw.githubusercontent.com/NuCivic/dkan/$DKAN_BRANCH/dkan-init.sh

# Make sure the download was at least successful.
if [ $? -ne 0 ] ; then
  echo ""
  echo "[Error] Failed to download the dkan-init.sh script from github dkan. Branch: $DKAN_BRANCH . Perhaps someone deleted the branch?"
  echo ""
  exit 1
fi

#Only stop on errors starting now..
set -e
# OK, run the script.
bash /tmp/dkan-init.sh $DKAN_MODULE $@ --skip-reinstall --branch=$DKAN_BRANCH
ahoy dkan module-link $DKAN_MODULE
ahoy dkan module-make $DKAN_MODULE
ahoy dkan reinstall
ahoy drush en $DKAN_MODULE -y

 #Fix for behat bug not recognizing symlinked feature files or files outside it's root. See https://jira.govdelivery.com/browse/CIVIC-1005
#cp dkan_workflow/test/features/dkan_workflow.feature dkan/test/features/.


recline.app.js
/**
 * @file
 * Provides options for recline visualization.
 */

(function ($) {
    var maxLabelWidth = 77;
    var labelMargin = 5;

    Drupal.behaviors.Recline = {
        attach: function (context) {
            var delimiter = Drupal.settings.recline.delimiter;
            var file = Drupal.settings.recline.file;
            var uuid = Drupal.settings.recline.uuid;
            var dkan = Drupal.settings.recline.dkan;
            var fileType = Drupal.settings.recline.fileType;
            var dataExplorerSettings = {
                grid: Drupal.settings.recline.grid,
                graph: Drupal.settings.recline.graph,
                map: Drupal.settings.recline.map
            };

            window.dataExplorer = null;
            window.explorerDiv = $('.data-explorer');

            // This is the very basic state collection.
            var state = recline.View.parseQueryString(decodeURIComponent(window.location.hash));
            if ('#map' in state) {
                state.currentView = 'map';
            } else if ('#graph' in state) {
                state.currentView = 'graph';
            } else if ('#timeline' in state) {
                state.currentView = 'timeline';
            }
            // Checks if dkan_datastore is installed.
            if (dkan || fileType == 'text/csv' || fileType == 'csv') {
                var drupal_base_path = Drupal.settings.basePath;
                var DKAN_API = drupal_base_path + 'api/action/datastore/search.json';
                var url = dkan ? (window.location.origin + DKAN_API + '?resource_id=' + uuid) : file;
                var options = dkan ? {} : {delimiter: delimiter};
                var ajax_options = {
                    type: 'GET',
                    url: dkan ? url : file,
                    dataType: dkan  ? 'json' : 'text',
                    beforeSend: function (jqXHR, settings) {
                        /* add url property and get value from settings (or from caturl)*/
                        jqXHR.dkan = dkan;
                    },
                    success: function(data, status, jqXHR) {
                        var dataset, views;
                        if (jqXHR.dkan) {
                            if ('success' in data && data.success) {
                                dataset = new recline.Model.Dataset({
                                    endpoint: window.location.origin + drupal_base_path + '/api',
                                    url: url,
                                    id: uuid,
                                    backend: 'ckan'
                                });
                                dataset.fetch();
                                views = createExplorer(dataset, state, dataExplorerSettings);
                            }
                            else {
                                $('.data-explorer').append('<div class="messages status">Error returned from datastore: ' + data + '.</div>');
                            }
                        }
                        else {
                            // Converts line endings in either format to unix format.
                            data = data.replace(/(\r\n|\n|\r)/gm,"\n");
                            dataset = new recline.Model.Dataset({
                                records: recline.Backend.CSV.parseCSV(data, options)
                            });
                            dataset.fetch();
                            views = createExplorer(dataset, state, dataExplorerSettings);
                            // The map needs to get redrawn when we are delivering from the ajax
                            // call.
                            $.each(views, function(i, view) {
                                if (view.id == 'map') {
                                    view.view.redraw('refresh');
                                }
                            });
                        }
                    }
                };
                if (dkan) {
                    ajax_options.error = function(data, status, jqXHR) {
                        $('.data-explorer').append('<div class="messages status">Unable to connect to the datastore.</div>');
                    };
                }
                else {
                    ajax_options.timeout = Drupal.settings.recline.ajax_timeout;
                    ajax_options.error = function(x, t, m) {
                        if (t === "timeout") {
                            $('.data-explorer').append('<div class="messages status">File was too large or unavailable for preview.</div>');
                        } else {
                            $('.data-explorer').append('<div class="messages status">Data preview unavailable.</div>');
                        }
                    };
                }
                $.ajax(ajax_options);
            }
            // Checks if xls.
            else if (fileType == 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet' || fileType == 'application/vnd.ms-excel') {
                var dataset = new recline.Model.Dataset({
                    url: file,
                    backend: 'dataproxy'
                });
                dataset.fetch();
                var views = createExplorer(dataset, state, dataExplorerSettings);
            }
            else {
                $('.data-explorer').append('<div class="messages status">File type ' + fileType + ' not supported for preview.</div>');
            }
        }
    };

    // make Explorer creation / initialization in a function so we can call it
    // again and again
    var createExplorer = function(dataset, state, settings) {
        // remove existing data explorer view
        var reload = false;
        if (window.dataExplorer) {
            window.dataExplorer.remove();
            reload = true;
        }
        window.dataExplorer = null;
        var $el = $('<div />');
        $el.appendTo(window.explorerDiv);

        var views = [];

        if (settings.grid) {
            views.push(
                {
                    id: 'grid',
                    label: Drupal.t('Grid'),
                    view: new recline.View.SlickGrid({
                        model: dataset
                    })
                }
            );
        }
        if (settings.graph) {
            state.graphOptions = {
                xaxis: {
                    tickFormatter:tickFormatter(dataset),
                },
                hooks:{
                    processOffset:[processOffset(dataset)],
                    bindEvents: [bindEvents],
                }
            };
            views.push(
                {
                    id: 'graph',
                    label: Drupal.t('Graph'),
                    view: new recline.View.Graph({
                        model: dataset,
                        state: state
                    })
                }
            );
        }
        if (settings.map) {
            views.push(
                {
                    id: 'map',
                    label: Drupal.t('Map'),
                    view: new recline.View.Map({
                        model: dataset
                    })
                }
            );
        }

        Drupal.settings.recline.args = {
            model: dataset,
            el: $el,
            state: state,
            views: views
        };

        // Getting base embed url.
        var urlBaseEmbed = $('.embed-code').text();
        var iframeOptions = {src: urlBaseEmbed, width:850, height:400};
        // Attaching router to dataexplorer state.
        window.dataExplorer = new recline.View.MultiView(Drupal.settings.recline.args);
        window.router = new recline.DeepLink.Router(window.dataExplorer);

        // Adding router listeners.
        changeEmbedCode = getEmbedCode(iframeOptions);
        window.router.on('init', changeEmbedCode);
        window.router.on('stateChange', changeEmbedCode);

        // Add map dependency just for map views.
        _.each(window.dataExplorer.pageViews, function(item, index){
            if(item.id && item.id === 'map'){
                var map = window.dataExplorer.pageViews[index].view.map;
                window.router.addDependency(new recline.DeepLink.Deps.Map(map, window.router));
            }
        });
        // Start to track state chages.
        window.router.start();

        $.event.trigger('createDataExplorer');
        return views;
    };

    $(".recline-embed a.embed-link").live('click', function(){
      $(this).parents('.recline-embed').find('.embed-code-wrapper').toggle();
      return false;
    });

    function getEmbedCode(options){
        return function(state){
            var iframeOptions = _.clone(options);
            var iframeTmpl = _.template('<iframe width="<%= width %>" height="<%= height %>" src="<%= src %>" frameborder="0"></iframe>');
            _.extend(iframeOptions, {src: iframeOptions.src + '#' + (state.serializedState || '')});
            var html = iframeTmpl(iframeOptions);
            $('.embed-code').text(html);
        };
    }

    function isInverted(){
        return dataExplorer.pageViews[1].view.state.attributes.graphType === 'bars';
    }

    function computeWidth (plot, labels) {
        var biggerLabel = '';
        for( var i = 0; i < labels.length; i++){
            if(labels[i].length > biggerLabel.length && !_.isUndefined(labels[i])){
                biggerLabel = labels[i];
            }
        }
        var canvas = plot.getCanvas();
        var ctx = canvas.getContext('2d');
        ctx.font = 'sans-serif smaller';
        return ctx.measureText(biggerLabel).width;
    }

    function resize (plot) {
        var itemWidth = computeWidth(plot, _.pluck(plot.getXAxes()[0].ticks, 'label'));
        var graph = dataExplorer.pageViews[1];
        if(!isInverted() && $('#prevent-label-overlapping').is(':checked')){
            var canvasWidth = Math.min(itemWidth + labelMargin, maxLabelWidth) * plot.getXAxes()[0].ticks.length;
            var canvasContainerWith = $('.panel.graph').parent().width();
            if(canvasWidth < canvasContainerWith){
                canvasWidth = canvasContainerWith;
            }
            $('.panel.graph').width(canvasWidth);
            $('.recline-flot').css({overflow:'auto'});
        }else{
            $('.recline-flot').css({overflow:'hidden'});
            $('.panel.graph').css({width: '100%'});
        }
        plot.resize();
        plot.setupGrid();
        plot.draw();
    }

    function bindEvents (plot, eventHolder) {
        var p = plot || dataExplorer.pageViews[1].view.plot;
        resize(p);
        setTimeout(addCheckbox, 0);
    }

    function processOffset (dataset) {
        return function(plot, offset) {
            if(dataExplorer.pageViews[1].view.xvaluesAreIndex){
                var series = plot.getData();
                for (var i = 0; i < series.length; i++) {
                      var numTicks = Math.min(dataset.records.length, 200);
                      var ticks = [];
                      for (var j = 0; j < dataset.records.length; j++) {
                        ticks.push(parseInt(j, 10));
                      }
                      if(isInverted()){
                        series[i].yaxis.options.ticks = ticks;
                      }else{
                        series[i].xaxis.options.ticks = ticks;
                      }
                }
            }
        };
    }

    function tickFormatter(dataset){
        return function (x) {
            x = parseInt(x, 10);
            try {
                if(isInverted()){
                    return x;
                }
                var field = dataExplorer.pageViews[1].view.state.get('group');
                var label = dataset.records.models[x].get(field) || "";
                if(!moment(String(label)).isValid() && !isNaN(parseInt(label, 10))){
                    label = parseInt(label, 10) - 1;
                }
                return label;
            } catch(e) {
                return x;
            }
        };
    }

    function addCheckbox() {
        $control = $('.form-stacked:visible').find('#prevent-label-overlapping');
        if(!$control.length){
            $form = $('.form-stacked');
            $checkboxDiv = $('<div class="checkbox"></div>').appendTo($form);
            $label = $('<label />', { 'for': 'prevent-label-overlapping', text: Drupal.t('Resize graph to prevent label overlapping') }).appendTo($checkboxDiv);
            $label.prepend($('<input />', { type: 'checkbox', id: 'prevent-label-overlapping', value: '' }));
            $control = $('#prevent-label-overlapping');
            $control.on('change', function(){
                resize(dataExplorer.pageViews[1].view.plot);
            });
        }
    }
})(jQuery);
